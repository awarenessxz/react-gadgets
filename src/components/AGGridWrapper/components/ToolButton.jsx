import React, { Fragment, useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import Badge from '@material-ui/core/Badge';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import IconButton from '@material-ui/core/IconButton';
import Grow from '@material-ui/core/Grow';
import Fade from '@material-ui/core/Fade';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import Tooltip from '@material-ui/core/Tooltip';

/**
 * Generic Button for Toolbar
 */
const ToolButton = props => {
    const [open, setOpen] = useState(false);
    const anchorRef = useRef(null);
    const prevOpen = useRef(open);

    // return focus to the button when we transitioned from !open -> open
    useEffect(() => {
        if (prevOpen.current === true && open === false) {
            anchorRef.current.focus();
        }
        prevOpen.current = open;
    }, [open]);

    const handleDropdownToggle = () => {
        setOpen(prevOpen => !prevOpen);
    };

    const handleCloseDropdown = e => {
        if (e) {
            if (anchorRef.current && anchorRef.current.contains(e.target)) {
                return;
            }
        }
        setOpen(false);
    };

    const handleMenuItemClick = menuItem => e => {
        props.onClick(menuItem);
        handleCloseDropdown(e);
    };

    const handleOnClickEventFromDropDownContent = params => {
        props.onClick(params);
        handleCloseDropdown();
    };

    const renderIcon = () => {
        const Icon = props.icon;
        if (props.disabledIcon) {
            const DisabledIcon = props.disabledIcon;
            return props.disable ? <DisabledIcon /> : <Icon />;
        } else {
            return <Icon />;
        }
    };

    const renderButtonWithBadge = () => {
        return (
            <Tooltip title={props.tooltipMsg}>
                <span>
                    <IconButton
                        id={props.buttonId}
                        aria-label={props.tooltipMsg}
                        onClick={props.onClick}
                        disabled={props.disable}>
                        <Badge badgeContent={props.badgeCount} color='primary'>
                            {renderIcon()}
                        </Badge>
                    </IconButton>
                </span>
            </Tooltip>
        );
    };

    const renderButtonWithMenu = () => {
        return (
            <Fragment>
                <Tooltip title={props.tooltipMsg}>
                    <IconButton
                        id={props.buttonId}
                        ref={anchorRef}
                        aria-controls={open ? 'menu-list-grow' : undefined}
                        aria-haspopup='true'
                        onClick={handleDropdownToggle}>
                        {renderIcon()}
                    </IconButton>
                </Tooltip>
                <Popper open={open} anchorEl={anchorRef.current} role={undefined} transition>
                    {({ TransitionProps, placement }) => (
                        <Grow
                            {...TransitionProps}
                            style={{
                                transformOrigin:
                                    placement === 'bottom' ? 'center top' : 'center bottom',
                            }}>
                            <Paper>
                                <ClickAwayListener onClickAway={handleCloseDropdown}>
                                    <MenuList autoFocusItem={open} id='menu-list-grow'>
                                        {props.menuItems.map((item, idx) => {
                                            return (
                                                <MenuItem
                                                    key={idx}
                                                    onClick={handleMenuItemClick(item)}>
                                                    {item}
                                                </MenuItem>
                                            );
                                        })}
                                    </MenuList>
                                </ClickAwayListener>
                            </Paper>
                        </Grow>
                    )}
                </Popper>
            </Fragment>
        );
    };

    const renderButtonWithDropdown = () => {
        const id = open ? 'transitions-popper' : undefined;
        return (
            <Fragment>
                <Tooltip title={props.tooltipMsg}>
                    <IconButton
                        id={props.buttonId}
                        ref={anchorRef}
                        aria-describedby={id}
                        aria-haspopup='true'
                        onClick={handleDropdownToggle}>
                        {renderIcon()}
                    </IconButton>
                </Tooltip>
                <Popper id={id} open={open} anchorEl={anchorRef.current} transition>
                    {({ TransitionProps }) => (
                        <Fade {...TransitionProps} timeout={350}>
                            <Paper elevation={3} style={{ padding: '20px', border: '1px solid' }}>
                                {React.cloneElement(props.dropdownContent, {
                                    onClick: handleOnClickEventFromDropDownContent,
                                })}
                            </Paper>
                        </Fade>
                    )}
                </Popper>
            </Fragment>
        );
    };

    const renderNormalButton = () => {
        return (
            <Tooltip title={props.tooltipMsg}>
                <IconButton
                    id={props.buttonId}
                    aria-label={props.tooltipMsg}
                    onClick={props.onClick}
                    disabled={props.disable}>
                    {renderIcon()}
                </IconButton>
            </Tooltip>
        );
    };

    const renderButton = () => {
        switch (props.buttonType) {
            case 'badge':
                return renderButtonWithBadge();
            case 'menu':
                return renderButtonWithMenu();
            case 'dropdown':
                return renderButtonWithDropdown();
            case 'normal':
            default:
                return renderNormalButton();
        }
    };

    return <Fragment>{renderButton()}</Fragment>;
};

ToolButton.defaultProps = {
    buttonType: 'normal',
    disable: false,
    dropdownContent: <span>Empty</span>,
};

ToolButton.propTypes = {
    /** tool tip message if there is any */
    tooltipMsg: PropTypes.string.isRequired,
    /** button id */
    buttonId: PropTypes.string.isRequired,
    /** icon */
    icon: PropTypes.object.isRequired,
    /** callback */
    onClick: PropTypes.func.isRequired,
    /** disable button */
    disable: PropTypes.bool,
    /** disabled icon */
    disabledIcon: PropTypes.object,
    /** Button Type */
    buttonType: PropTypes.oneOf(['normal', 'badge', 'menu', 'dropdown']),
    /** badge counter (is Required when buttonType = "badge") */
    badgeCount: PropTypes.number,
    /** menu items (is Required when buttonType = "menu") */
    menuItems: PropTypes.arrayOf(PropTypes.string),
    /** dropdown Content -- Eg. <ReactComponent /> (is Required when buttonType = "dropdown") */
    dropdownContent: PropTypes.object,
};

export default ToolButton;
