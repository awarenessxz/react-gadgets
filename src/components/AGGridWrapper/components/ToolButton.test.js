import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import ToolButton from './ToolButton';
import GetApp from '@material-ui/icons/GetApp';

// function to render component before each test
const renderComponent = props => {
    const defaultProps = {
        tooltipMsg: 'Tooltip Message',
        icon: GetApp, // for mocking only (just a random icon in material-ui)
        onClick: jest.fn(),
    };
    const merged = { ...defaultProps, ...props };
    return render(<ToolButton {...merged} />); // returns methods (refer to api)
};

// 1. Testing if component renders properly
describe('Testing if component renders properly', () => {
    // Snapshot Testing
    it('Snapshot Testing', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    // Testing Dom Elements: check if renders with correct value (Note the usage of data-testid in component)
    describe('component renders with correct state (Testing Dom Elements)', () => {
        it('buttonType="normal" renders properly', () => {
            const toolTip = 'normal button';
            renderComponent({ buttonType: 'normal', tooltipMsg: toolTip });
            expect(screen.getByRole('button', { name: toolTip })).toBeTruthy();
        });

        it('buttonType="badge" renders properly', () => {
            const toolTip = 'badge button';
            const badgeCount = 5;
            renderComponent({ buttonType: 'badge', tooltipMsg: toolTip, badgeCount: badgeCount });
            expect(screen.getByRole('button', { name: toolTip })).toBeTruthy();
            expect(screen.getByText(badgeCount.toString())).toBeTruthy();
        });

        it('buttonType="menu" renders properly', () => {
            const toolTip = 'menu button';
            const menuItems = ['menu 1', 'menu 2'];
            renderComponent({ buttonType: 'menu', tooltipMsg: toolTip, menuItems: menuItems });
            expect(screen.getByRole('button', { name: toolTip })).toBeTruthy();
        });

        it('buttonType="dropdown" renders properly', () => {
            const toolTip = 'dropdown button';
            const clickFn = jest.fn();
            const MockComponent = () => <div>Hello World</div>;
            renderComponent({
                buttonType: 'dropdown',
                tooltipMsg: toolTip,
                dropdownContent: <MockComponent />,
                onClick: clickFn,
            });
            expect(screen.getByRole('button', { name: toolTip })).toBeTruthy();
            // opens dropdown
            fireEvent.click(screen.getByRole('button', { name: toolTip }));
            expect(screen.queryByText('Hello World')).toBeInTheDocument();
        });

        describe('disabled button', () => {
            it('renders properly', () => {
                const toolTip = 'disabled button';
                renderComponent({ disabled: true, tooltipMsg: toolTip });
                expect(screen.getByRole('button', { name: toolTip })).toBeDisabled();
            });

            it('with different disabled icon renders properly', () => {
                const toolTip = 'disabled button';
                renderComponent({ disabled: true, disabledIcon: GetApp, tooltipMsg: toolTip });
                expect(screen.getByRole('button', { name: toolTip })).toBeDisabled();
            });
        });
    });
});

// 2. Testing Events (Eg. onClick)
describe('Testing Events', () => {
    describe('onClick is called', () => {
        it('when ToolButton[type="normal"] is clicked', () => {
            const toolTip = 'normal button';
            const clickFn = jest.fn();
            renderComponent({
                buttonType: 'normal',
                tooltipMsg: toolTip,
                onClick: clickFn,
            });
            fireEvent.click(screen.getByRole('button', { name: toolTip }));
            expect(clickFn).toHaveBeenCalled();
        });

        it('when ToolButton[type="badge"] is clicked', () => {
            const toolTip = 'badge button';
            const clickFn = jest.fn();
            renderComponent({
                buttonType: 'badge',
                tooltipMsg: toolTip,
                badgeCount: 5,
                onClick: clickFn,
            });
            fireEvent.click(screen.getByRole('button', { name: toolTip }));
            expect(clickFn).toHaveBeenCalled();
        });

        it('when menu item in ToolButton[type="menu"] is clicked', () => {
            const toolTip = 'menu button';
            const menuItems = ['menu 1', 'menu 2'];
            const clickFn = jest.fn();
            renderComponent({
                buttonType: 'menu',
                tooltipMsg: toolTip,
                menuItems: menuItems,
                onClick: clickFn,
            });
            // opens menu
            fireEvent.click(screen.getByRole('button', { name: toolTip }));
            // click on menu item
            fireEvent.click(screen.getByRole('menuitem', { name: menuItems[0] }));
            expect(clickFn).toHaveBeenCalled();
        });

        it('when onClick is called in ToolButton[type="dropdown"] is clicked', () => {
            const toolTip = 'dropdown button';
            const clickFn = jest.fn();
            const MockComponent = props => <button onClick={props.onClick}>Hello World</button>;
            renderComponent({
                buttonType: 'dropdown',
                tooltipMsg: toolTip,
                dropdownContent: <MockComponent />,
                onClick: clickFn,
            });
            // opens dropdown
            fireEvent.click(screen.getByRole('button', { name: toolTip }));
            // trigger click
            fireEvent.click(screen.getByRole('button', { name: 'Hello World' }));
            expect(clickFn).toHaveBeenCalled();
        });
    });
});
