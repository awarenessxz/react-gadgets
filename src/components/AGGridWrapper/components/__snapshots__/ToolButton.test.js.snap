// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing if component renders properly Snapshot Testing 1`] = `
<DocumentFragment>
  <span
    class=""
    title="Tooltip Message"
  >
    <button
      aria-label="Tooltip Message"
      class="MuiButtonBase-root MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </span>
</DocumentFragment>
`;
